challenges/cloud/sre.md

```diff
@@ -0,0 +1,376 @@
+checkout the repository and execute the tests themselves.
+
+Do not try to achieve full test coverage. This will take too long. Write enough
+to exercise the different key components and show they are working as
+intended, as well as to demonstrate your approach to automation.
+
+## Deployment
+
+Levels 3+ include a deployment step which should deploy the solution to a local
+Kubernetes cluster. The choice of which Kubernetes cluster is up to you, but 
+please choose one that has the ability to run on macOS and Linux.
+
+## Key Dependencies
+
+You are welcome to solve this challenge using tools you are most familiar with,
+but, at a minimum, please include the following:
+
+* Go
+* make
+* Docker
+
+## Level 1
+
+### Server
+
+* HTTP API to retrieve the replica count of the Kubernetes Deployment
+
+### Automation
+
+* Ability to build the server in a Docker container
+* Ability to run the server
+
+## Level 2
+
+### Server
+
+* HTTP API to retrieve the replica count of the Kubernetes Deployment
+* HTTP API to set the replica count of the Kubernetes Deployment
+
+### Automation
+
+* Ability to build the server in a Docker container
+* Ability to run the server
+* Ability to execute integration tests against the local Kubernetes cluster
+
+## Level 3
+
+### Server
+
+* HTTP API to retrieve the replica count of the Kubernetes Deployment
+* HTTP API to set the replica count of a the Kubernetes Deployment
+* HTTP API to get the list of available Deployments in the Kubernetes cluster
+* HTTP health check verifying Kubernetes connectivity
+
+### Automation
+
+* Ability to build the server in a Docker container
+* Ability to run the server
+* Ability to execute integration tests against the local Kubernetes cluster
+* GitHub Actions for every commit that verifies code quality, ensures the software builds, and all tests pass
+
+### Deployment
+
+* Create a helm chart for the service
+
+## Level 4
+
+### Design
+
+In addition to including details about the requirements below, the design
+document should include a section describing how this solution could be deployed
+to an AWS EKS cluster.
+
+### Server
+
+* HTTP API to retrieve the replica count of the Kubernetes Deployment
+* HTTP API to set the replica count of a the Kubernetes Deployment
+* HTTP API to get the list of available Deployments in the Kubernetes cluster
+* HTTP health check verifying Kubernetes connectivity
+* Extend the API to support a difference between current and desired state
+* Extend the Server to support the ability to show the differences between current and desired state
+* Secure connections with mTLS
+
+### Automation
+
+* Ability to build the server in a Docker container
+* Ability to run the server
+* Ability to execute integration tests against the local Kubernetes cluster
+* GitHub Actions for every commit that verifies code quality, ensures the software builds, and all tests pass
+* Produce production ready releases (binaries and docker image)
```

Comment:
This is pretty subjective and can mean different things to difference candidates depending on their background. Is there any way we can make this more fair by listing the specific things we'd like to see?

Reply:
> This is pretty subjective and can mean different things to difference candidates depending on their background.

This is a feature, not a bug

> Is there any way we can make this more fair by listing the specific things we'd like to see?

We could. We could also field those questions in slack, or the candidate can make assumptions. In any case, how the candidate interprets what it means to be production ready is an interesting signal to me. I'd rather keep this ambiguous so I can see what candidates come up with in the given time constraints rather than be explicit

Reply:
I'll leave this up to you all then, but I'm not sure I agree.

Production ready to someone who's coming from Google is going to mean something different to someone who is maybe newer in their career, has a non-traditional background, or hasn't had the opportunity to work in big tech.

At the end of the day, both people are capable of doing the same work, I want to hire them both, not just the one who had the privilege to work somewhere where "production ready" more closely aligned with what our views.

Reply:
Yeah, I think we're going to field some questions about this. We'll leave it in for now and update once we get a better idea of how to communicate without ambiguity.